src/compute/src/render/mod.rs

```diff
@@ -250,11 +250,26 @@ fn intercept_source_instantiation_frontiers<G>(
 where
     G: Scope<Timestamp = mz_repr::Timestamp>,
 {
+    let mut previous_time = None;
     source_instantiation.inspect_container(move |event| {
         if let Err(frontier) = event {
-            if let Some(time) = frontier.get(0) {
-                for dataflow_id in dataflow_ids.iter() {
-                    logger.log(ComputeEvent::SourceFrontier(*dataflow_id, source_id, *time));
+            for dataflow_id in dataflow_ids.iter() {
+                if let Some(previous) = previous_time {
+                    logger.log(ComputeEvent::SourceFrontier(
+                        *dataflow_id,
+                        source_id,
+                        previous,
+                        -1,
+                    ));
+                }
+                if let Some(time) = frontier.get(0) {
+                    logger.log(ComputeEvent::SourceFrontier(
+                        *dataflow_id,
+                        source_id,
+                        *time,
+                        1,
+                    ));
+                    previous_time = Some(*time);
```

Comment:
It seems the `for dataflow_id in dataflow_ids` loop should be inside the `previous_time` check, and inside the `frontier.get(0)` block. Otherwise, we'll only emit correct progress information for the first dataflow, but not subsequent ones.

Reply:
You are absolutely right. I was too trigger happy here with hoisting the for loops across the ifs. To try to strengthen the code even a bit more, I also added handling for the case of double-reporting of the empty frontier, since I was not sure that Timely gives us the contract that an empty frontier would be reported only once.

Reply:
Timely should only report the empty frontier once, because afterwards, no more data or progress can be observed. If it reports it multiple times, then this is surprising and we should fix it. However, it's great to to be defensive when implementing the operator.